### DISCUSSION TITLE
Embedded images

### DISCUSSION THREAD

**Discussion by HernanFdez** [resend/react-email#1281] [Ideas]
### Goals

1. Capability for embedding an image directly into the produced email, instead of by specifying a remote URL src attribute, so that it does not need to be retrieved by the mail viewer from a remote hosting service at view time.

### Background

Embedding images directly in emails, I think have some advantage over writning down just a reference to a remote web resource, since it does not need to be externally retrieved by the mail viewer every time. This means it does not need to be kept online indefinitely just in case some day someone wants to view a given email.

First solution that came to mind was simply using a Img tag of the form `<Img src="data:image/_format_;base64,_content_">`, where `_content_` is the base64 string representation of the image.

But then I found out that, at least Gmail [(but probably other mail viewers too)](https://www.caniemail.com/features/image-base64/), does not support this, and instead expects an embedded image to be included as a part of a multipart message. So this does not work.

### Proposal

I do not really have a profound suggestion about how to implement this.
I am not even sure if this issue should belong to react-email or resend-node.

Anyhow, what I would like to have as the final result is simply some way to create an email with react-email as a react component contating `<Img />` tags with `src="data:image/_format_;base64,_content_"` and send it with `resend-node`, so that the images are embedded in the resulting message using the multipart format that Gmail expects.

**Comment by gabrielmfern** [comment]
> and instead expects an embedded image to be included as a part of a multipart message

Do you mean attaching the image and then referencing it on the HTML?

**Reply by HernanFdez** [reply]
Is there already a way to do that? Could you explain how exactly? 
Apologies for putting it as a feature request if that is the case.

**Reply by gabrielmfern** [reply]
It's just that I've never had a need myself to show attachments of the email on its HTML so I'm not sure if that's possible, 
just wanted to see if that's what you meant here.

Answering your question, nope, we don't really have a way to automatically add the attachment. But the Resend SDK does support attachments https://github.com/resend/resend-examples/tree/main/with-attachments

**Reply by gabrielmfern** [reply]
Let me rephrase my question: Do you mean that Gmail expects that the image is attached and then referenced inside of the HTML?

**Reply by gabrielmfern** [reply]
Would be awesome if you could link resources you found about this as well. 

Been thinking about something like this for a while, it would actually be really a great improvement on DX if we could automatically attach and use the content inside of the email HTML to allow for stuff like Plots, QR Codes and other kinds of complex things that would generally require something like https://quickchart.io/documentation/send-charts-in-email/

**Reply by HernanFdez** [reply]
Hi, take a look at this [SendGrid docs](https://sendgrid.com/en-us/blog/embedding-images-emails-facts#:~:text=CID%20images%20work%20by%20attaching,when%20the%20user%20opens%20it.) if you may.

As far as I can see, the one thing that's missing is a `cid` key for the `attachments` items in `resend.emails.send()` so that, when set to `some_id`, sets a `Content-ID: <some_id>` header for the part of the multipart message that contains the image data.
Currently it only sets these three headers:

`Content-Type: image/jpeg; name=some_name.jpg`
`Content-Transfer-Encoding: base64`
`Content-Disposition: attachment; filename=some_name.jpg`

But, contrary to what I would have guessed, the `name` cannot be used for reference in the html, so we do need the `Content-ID` for this purpose.

With that as an option, I think one could just include an image as attachment when calling `resend.emails.send()`, specifying `cid: some_id`, and then reference that in the source of an image in the html (or React) code like `<img src="cid:some_id"/>`

**Reply by gabrielmfern** [reply]
Ok, now it's much clearer. This is a `resend-node` missing feature, just opened an issue for it https://github.com/resend/resend-node/issues/319.

Would be really cool for us to do this automatically with react email though, but we can't really make it depend on Resend's SDK.
Thinking of a proposal of an API for this that would be flexible about what you are using to send the email:

```jsx
const { Image: MyImage, attachmentMetadata } = createImageAttachment({ 
  sendingWith: 'resend', // the `attachmentMetadata` would change its type and values according to this
  path: '/path/to/image' 
}); // it could read the file and return the content on the metadata, maybe as base64?

// so this can be used like
export default function MyEmail() {
  return <MyImage/>;
}
```

And then can be added into the attachments using the `attachmentMetadata` that would fit the API for the whatever they may be using to send the email

---

As a side note, when looking at it from their article it seems using attached images like this doesn't have the best support either
so it might not be the way to go for the things I mentioned.

**Comment by RodolfoSilva** [comment]
Related issues:

https://github.com/resend/resend-django-smtp-example/issues/2

https://github.com/resend/resend-node/issues/319

---

This feature is also missing in the Resend API.

https://resend.com/docs/api-reference/emails/send-email
